drivers/ata/ahciem.c

```diff
@@ -0,0 +1,588 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer,
+ *    without modification, immediately at the beginning of the file.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/spinlock.h>
+#include <scsi/scsi.h>
+#include <scsi/scsi_host.h>
+#include <scsi/scsi_cmnd.h>
+#include <scsi/scsi_device.h>
+#include <scsi/scsi_transport.h>
+#include <scsi/scsi_proto.h>
+#include <target/target_core_base.h>
+#include <linux/libata.h>
+#include <linux/enclosure.h>
+#include "ahci.h"
+
+#define DRV_NAME	"ahciem"
+#define DRV_VERSION	"0.1"
+
+#define AHCIEM_RBUF_SIZE	576	/* size from libata-scsi.c */
+
+static DEFINE_SPINLOCK(ahciem_rbuf_lock);
+static u8 ahciem_rbuf[AHCIEM_RBUF_SIZE];
+
+struct ahciem_enclosure {
+	struct ata_host *host;
+	u8 status[AHCI_MAX_PORTS][4];
+};
+
+struct ahciem_args {
+	struct scsi_cmnd *cmd;
+	struct ahciem_enclosure *enc;
+};
+
+/*
+ * Utility functions
+ */
+
+static __inline void scsi_ulto2b(u32 val, u8 *bytes)
+{
+	bytes[0] = (val >> 8) & 0xff;
+	bytes[1] = val & 0xff;
+}
+
+static __inline void scsi_ulto4b(u32 val, u8 *bytes)
+{
+	bytes[0] = (val >> 24) & 0xff;
+	bytes[1] = (val >> 16) & 0xff;
+	bytes[2] = (val >> 8) & 0xff;
+	bytes[3] = val & 0xff;
+}
+
+static void ahciem_rbuf_fill(struct ahciem_args *args,
+		unsigned int (*actor)(struct ahciem_args *args, u8 *rbuf))
+{
+	unsigned int rc;
+	struct scsi_cmnd *cmd = args->cmd;
+	unsigned long flags;
+
+	spin_lock_irqsave(&ahciem_rbuf_lock, flags);
+
+	memset(ahciem_rbuf, 0, AHCIEM_RBUF_SIZE);
+	rc = actor(args, ahciem_rbuf);
+	if (rc == 0)
+		sg_copy_from_buffer(scsi_sglist(cmd), scsi_sg_count(cmd),
+				    ahciem_rbuf, AHCIEM_RBUF_SIZE);
+
+	spin_unlock_irqrestore(&ahciem_rbuf_lock, flags);
+
+	if (rc == 0)
+		cmd->result = SAM_STAT_GOOD;
+}
```

Comment:
I don't see you ever use ALLOCATION LENGTH field of the commands to know how much data is requested, neither you track how much data you have.  If initiator requested more data than you have, you should probably somehow return the residual amount.  In some cases initiator may need to know how much data in the received buffer are valid.  Your present code may work, but it is not completely correct.

Reply:
It works the same way in libata-scsi.c so I'm not sure it's worth correcting.